### ISSUE TITLE
Sign commits created by Batch Changes with GitHub App (Tracking)

### ISSUE THREAD

**Issue by BolajiOlajide** [sourcegraph/sourcegraph#48611]
This issue is related to #15271. 

We have requests from some of our customers to add an extra layer of security when creating changesets from a Batch Change to have the associated commits in a changeset signed by the author of the Batch Change. To achieve tho

* [Gitlab](https://docs.gitlab.com/ee/user/project/repository/ssh_signed_commits/)
* [Github](https://github.blog/changelog/2022-08-23-ssh-commit-verification-now-supported/)
* Bitbucket [(Unsupported)](https://community.atlassian.com/t5/Bitbucket-questions/Sign-commit-with-SSH-Key/qaq-p/2002101)
* Azure DevOps [(Unsupported)](https://developercommunity.visualstudio.com/t/show-whether-commits-are-verified-with-gpg-key-on/383281)

The proposal here is to have users generate SSH keys on Sourcegraph, the private key will be encrypted and stored and used for creating commits, while the public key will be displayed to the user to be added to their code hosts.

Thanks to Erik, we have a draft PR [here](https://github.com/sourcegraph/sourcegraph/pull/43286/files). 

The scope of work involved to finish this implementation includes:
- Extending the user credential and site admin credential types by a flag "I want signing bool" that can be toggled from the UI
  - If that is checked, we will show the generated SSH key regardless of if ssh on the code host is enabled
  - And inform the user that they need to store that key in their profile as a signing key
- Polishing the gitserver part of that PR, likely want some testing and such
- Adding some feature gating because older GitHub instances don't support this signing type
- Investigating if GitLab supports this, implementing support for it if so
- Documenting
- Testing

### Tracked issues

@unassigned

- [ ] https://github.com/sourcegraph/sourcegraph/issues/51219 
- [ ] https://github.com/sourcegraph/sourcegraph/issues/52792 
- [ ] https://github.com/sourcegraph/sourcegraph/issues/53320 

Completed
- [x] (🏁 20 days ago) https://github.com/sourcegraph/sourcegraph/issues/50940 
- [x] (🏁 20 days ago) https://github.com/sourcegraph/sourcegraph/issues/50944 

@BolajiOlajide

- [ ] https://github.com/sourcegraph/sourcegraph/pull/52333 

Completed
- [x] (🏁 22 days ago) https://github.com/sourcegraph/sourcegraph/issues/52264 
- [x] (🏁 19 days ago) https://github.com/sourcegraph/sourcegraph/issues/52270 
- [x] (🏁 19 days ago) https://github.com/sourcegraph/sourcegraph/issues/52324 
- [x] (🏁 19 days ago) https://github.com/sourcegraph/sourcegraph/issues/52346 
- [x] (🏁 12 days ago) https://github.com/sourcegraph/sourcegraph/issues/52494 
- [x] (🏁 2 days ago) https://github.com/sourcegraph/sourcegraph/issues/52475 (PRs: ~[#52979](https://github.com/sourcegraph/sourcegraph/pull/52979)~)

@courier-new

Completed
- [x] (🏁 19 days ago) https://github.com/sourcegraph/sourcegraph/issues/52342 (PRs: ~[#52430](https://github.com/sourcegraph/sourcegraph/pull/52430)~)
- [x] (🏁 11 days ago) https://github.com/sourcegraph/sourcegraph/issues/52488 (PRs: ~[#52749](https://github.com/sourcegraph/sourcegraph/pull/52749)~)
- [x] (🏁 8 days ago) https://github.com/sourcegraph/sourcegraph/issues/52493 (PRs: ~[#52492](https://github.com/sourcegraph/sourcegraph/pull/52492)~)
- [x] (🏁 7 days ago) https://github.com/sourcegraph/sourcegraph/pull/52907 
- [x] (🏁 5 days ago) https://github.com/sourcegraph/sourcegraph/issues/52341 
- [x] (🏁 5 days ago) https://github.com/sourcegraph/sourcegraph/issues/53032 
- [x] (🏁 2 days ago) https://github.com/sourcegraph/sourcegraph/pull/53286 
- [x] (🏁 2 days ago) https://github.com/sourcegraph/sourcegraph/issues/52220 (PRs: ~[#52979](https://github.com/sourcegraph/sourcegraph/pull/52979)~)
- [x] (🏁 1 day ago) https://github.com/sourcegraph/sourcegraph/issues/53280 (PRs: ~[#53212](https://github.com/sourcegraph/sourcegraph/pull/53212)~)

@st0nebraker

- [ ] https://github.com/sourcegraph/sourcegraph/issues/52224 

Completed
- [x] (🏁 25 days ago) https://github.com/sourcegraph/sourcegraph/issues/50935 
- [x] (🏁 25 days ago) https://github.com/sourcegraph/sourcegraph/issues/51317 
- [x] (🏁 22 days ago) https://github.com/sourcegraph/sourcegraph/issues/52264 
- [x] (🏁 19 days ago) https://github.com/sourcegraph/sourcegraph/issues/52270 
- [x] (🏁 19 days ago) https://github.com/sourcegraph/sourcegraph/issues/52324 
- [x] (🏁 19 days ago) https://github.com/sourcegraph/sourcegraph/issues/52465 
- [x] (🏁 8 days ago) https://github.com/sourcegraph/sourcegraph/issues/52484 
- [x] (🏁 7 days ago) https://github.com/sourcegraph/sourcegraph/issues/52223 (PRs: ~[#52979](https://github.com/sourcegraph/sourcegraph/pull/52979)~)

Stretch goal:
- [ ] #51431

**Comment by courier-new** [comment]
Given what we've been discussing about https://github.com/sourcegraph/sourcegraph/issues/32196, it sounds like that'll probably come first, then at least some initial feedback gathering, before we would consider prioritizing this. So maybe for now, backlog?

**Comment by courier-new** [comment]
We resolved in our last team sync that this would likely require minimal additional effort to complete over https://github.com/sourcegraph/sourcegraph/issues/32196 but would nontrivially improve the UX, so we will be planning to move forward with this issue for the next round of Job Fair work.

**Comment by courier-new** [comment]
I saw during sprint planning there was another ticket with an open investigation as to where we store the commit signing keys. I can't seem to find that ticket now. Did we come to a decision about that?

**Comment by st0nebraker** [comment]
I'm unclear on a few things in these project specs:
(1) 
> Extending the user credential and site admin credential types by a flag "I want signing bool" that can be toggled from the UI
> If that is checked, we will show the generated SSH key regardless of if ssh on the code host is enabled
> And inform the user that they need to store that key in their profile as a signing key
Do we need a design for this? I'm not familiar with where this admin toggle would live.

(2) Will the SSH signing key be able to be used universally for GitLab (if GitLab supports this) and GitHub ? Or would we need one for each code host? This will effect the design and how we display the info and give more actions here for the user.

(3) On a similar note, for #50944, if we extend the [`CreateBatchChangesCredential`](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/frontend/internal/batches/resolvers/resolver.go?L1116) mutation to generate this SSH key, how would this work with the current variables we require in the request? `externalServiceKind` & `externalServiceURL` _could_ be static if this feature is only supported in GitHub, or we can give some user action to determine if it's for GitHub/GitLab. But `credential` is also required, but we're trying to generate one here, correct? Not store what a different code host generates. Would it make more sense to create a new endpoint to generate the SSH signing key? I'm just trying to understand the shape of the data & request the client needs to make as I work on #50935. @sourcegraph/batchers 

<img width="361" alt="image" src="https://user-images.githubusercontent.com/59381432/234347290-8d545347-3f5e-4a7a-bdba-b41759d01975.png">

**Comment by courier-new** [comment]
So I don't want to speak for Bolaji, but in the hopes that this might at least partially unblock you, here's what I would personally interpret/suggest for these things:

> 1. Do we need a design for this? I'm not familiar with where this admin toggle would live.

I'm actually really glad you asked about this because I totally missed this part of the project spec, and I think it cleared up one of my own questions! 😂 I believe this admin toggle would belong on the "add credential" modal for configuring code host credentials on that same user/admin settings page.

cc @danielmarquespt for thoughts on this, but here's what I might propose in terms of design:

1. We show a checkbox at the end of the "add credential" modal form to enable a user to toggle commit signing on/off for that code host credential:
    <img width="551" alt="image" src="https://user-images.githubusercontent.com/8942601/234357760-81603f59-ab4e-4b0d-a96d-f1acdb8d8de7.png">

2. When toggled, we expand the form to show a region similar to ["Step 2: Get SSH Key"](https://storybook.sgdev.org/?path=/story/web-batches-settings-addcredentialmodal--requires-ss-hstep-2) on the version of the modal where the code host has SSH enabled for the connection, except maybe the text reads like this:
    <img width="535" alt="image" src="https://user-images.githubusercontent.com/8942601/234359395-b0f8b756-ef65-4f20-a566-7fd44caec74b.png">

3. The "Configuration instructions" link should link to the docs pages referenced in the initial issue description above for the appropriate code host. 

4. If the code host type does not support commit signing, e.g. Bitbuckets or ADO, we disable the checkbox with a tooltip:
    <img width="554" alt="image" src="https://user-images.githubusercontent.com/8942601/234360544-f9652fcb-a83a-4717-91ac-152bb6d8744e.png">

    This should match the style of the execution options dropdown on the spec editor page:
    <img width="322" alt="image" src="https://user-images.githubusercontent.com/8942601/234360642-b1d2f154-0ec1-4f01-9cb2-38c8b6f85a6f.png">

    The checkbox could also be disabled if the user has not yet created any key to sign with:
    <img width="536" alt="image" src="https://user-images.githubusercontent.com/8942601/234364003-61091073-2d5f-4185-8e51-f24b82cbf0ec.png">


5. I expect we will add another arg to the `CreateBatchChangesCredential` mutation to reflect whether or not this checkbox is toggled, such as `preferSignedCommits: Boolean`, so that we can use that information when acting with this credential.

6. Once successfully configured, the credential node should show an indication that is uses commit signing, such as with an inline badge:
    <img width="985" alt="image" src="https://user-images.githubusercontent.com/8942601/234366840-999cebf2-1b41-4073-98c3-161b6eceab35.png">

    Or likewise if it's from a global token:
    <img width="974" alt="image" src="https://user-images.githubusercontent.com/8942601/234366954-dc700590-49eb-432f-8b8f-fc017ed79f0f.png">


8. Lastly, for code hosts with SSH enabled, you can view credentials after the fact, if you forgot the public key. We should mirror the form checkbox here in a readonly mode:
    <img width="542" alt="image" src="https://user-images.githubusercontent.com/8942601/234365487-4f6fdb7f-6598-40c8-828a-d24a49672b0e.png">


---

> 2. Will the SSH signing key be able to be used universally for GitLab (if GitLab supports this) and GitHub ? Or would we need one for each code host? This will effect the design and how we display the info and give more actions here for the user.

My current understanding is that a user generates a single signing key to use across all of their code hosts that would support it. The way it's presented in the design screenshot seems to support that. We definitely ought to convey this appropriately in the modal when that appears when you generate it.

---

> 3. On a similar note, for https://github.com/sourcegraph/sourcegraph/issues/50944, if we extend the `CreateBatchChangesCredential` mutation to generate this SSH key, how would this work with the current variables we require in the request?

This is exactly the question I asked on https://github.com/sourcegraph/sourcegraph/issues/50944 myself. 🙂 So I definitely want to hear what Bolaji had in mind for this. My personal feeling is that it ought to be a separate mutation and write the key to a separate table, but for now to unblock you on that I'd advise just skipping the onclick handler for generating the new key, if you can.

**Comment by st0nebraker** [comment]
Whoops, sorry I missed your same question on #50944 - or my page wasn't refreshed @courier-new !

Thank you for the quick mockups, it looks great! I like this path a lot. I have a few questions on this suggested flow, we can iron out in our sync tomorrow. I have enough to work on until then 🙂

- What if a code host credential is already setup but the user wants to add a SSH signing key? We would need some sort of view/edit button to bring up that modal again, to land on this view again. Does that sound right to you?
![image](https://user-images.githubusercontent.com/59381432/234396361-cadb4c2a-ffdd-4723-ad73-62cd7b35a76c.png)

> When toggled, we expand the form to show a region similar to ["Step 2: Get SSH Key"](https://storybook.sgdev.org/?path=/story/web-batches-settings-addcredentialmodal--requires-ss-hstep-2) on the version of the modal where the code host has SSH enabled for the connection, except maybe the text reads like this:
![image](https://user-images.githubusercontent.com/59381432/234396646-e5aae7a9-a0da-4a36-843e-f112d38c53c5.png)

- To make sure I'm clear on order of operations, it sounds like the token generation should occur separately or after the code host credentials are made. What if the PAT is invalid? We shouldn't generate an SSH sign key until after the credentials are made, correct? Or at the same time, in which case we'd need to show the key after the modal submission.

> The checkbox could also be disabled if the user has not yet created any key to sign with:
![image](https://user-images.githubusercontent.com/59381432/234398737-d091fe8a-d4fc-4333-bdf3-22b5bb7da3ec.png)

- Do you mean if there is no universal SSH sign key setup? Let me make sure I'm clear on combining the design below with your suggested flow. So there would be a universal SSH sign key and a separate one for each supported code host? Or what we display in the code host modal is just a replica of the universal SSH key?
![image](https://user-images.githubusercontent.com/59381432/234398891-68be30ab-fa0b-4f96-b23b-7180da2a23d2.png)

**Comment by danielmarquespt** [comment]
@courier-new you are hired as a designer! I think your solution works great.

I do have a question, where do users setup their private keys? Is this something Sourcegraph allows? Im wondering if on thisemessage we could help the user by telling them where to add the private keys.

![Screenshot 2023-04-26 at 13 09 29](https://user-images.githubusercontent.com/7814431/234570305-175d32e8-0e87-4954-b1b9-820bb15497de.png)

Another option to address @st0nebraker comments about validation of the two things is two have a separate action that triggers a modal to handle all the commit signing things, that would display only if the code-host has that option enabled, and disabled (with the tooltip for the others)

![Screenshot 2023-04-26 at 13 14 06](https://user-images.githubusercontent.com/7814431/234571303-ba999c3a-270c-48a1-8004-f1c161bc75f3.png)

**Comment by courier-new** [comment]
We answered Becca's and Daniel's questions and more today in our sync on this and have iterated upon this workflow once more [here](https://github.com/sourcegraph/sourcegraph/issues/50935#issuecomment-1523908984)! The biggest confusion about where the key comes from and what if it's not already created has been totally sidestepped because we determined we could just reuse the same SSH key that we generate to authenticate to code hosts that require one for auth! This also simplified the UI a bit since we could consolidate to a single public key displayed on Step 2 of the form. 😄
